core/blockchain.go

```diff
@@ -996,6 +996,8 @@ func (bc *BlockChain) Stop() {
 	bc.scope.Close()
 	close(bc.quit)
 	bc.StopInsert()
+	bc.chainmu.Lock()
```

Comment:
I think we should take the lock before calling `StopInsert`. It's a little bit safer.

Reply:
I don't think so

When we're procesing (maybe a couple of hundred blocks), in `insertChain`, we're holding the mutex and occasionally checking: 
```
		if bc.insertStopped() {
			log.Debug("Abort during block processing")
			break
		}
```
If we grab the chainMu first, then we'll always wait out the current batch of insertions when we try to stop.